design/expected-hardware-configuration-validation.md

```diff
@@ -0,0 +1,262 @@
+                    version:
+                        RAID:
+                            type: String
+                        BasebandManagement:
+                            type: String
+                        BIOS:
+                            type: String
+                        IDRAC:
+                            type: String
+                type: Object
+            required:
+            - customFilter
+            - namespace
+            - profileName
+        rules:
+            properties:
+                feild:
+                    type: String
+                operation:
+                    type: String
+                value:
+                    type: String
+        type: object
+        required:
+        - feild
+        - operation
+        - value
+    ```
+* Write a new API as hardware-classification-controller/api/v1alpha1 and new kind(CRD) HardwareClassificationController.
+    e.g.
+        
+       kubebuilder create api --group metal3.io --version v1alpha1 --kind HardwareClassificationController
+    
+    - This will create the files api/v1alpha1/hardwareClassificationController_types.go where the API is defined and the controller/hardwareClassificationController_controller.go where the reconciliation business logic is implemented for this Kind(CRD).
+
+    - Implement a new function fetchHost() which will fetch all baremetal hosts from Baremetal-Operator of namespace 'metal3' provided by user in CRD_yaml. This function will return a list of filtered hosts(hosts in status 'ready' or 'inspecting'). 
+
+    - In hardwareClassificationController_controller.go, reconcile function will call fetchHost()        function to fetch all baremetal hosts and also extract expectedHardwareConfiguration or expression rules from `metal3.io_HardwareClassificationController_crd.yaml`.
+
+    - Create a classification_manager.go file. If user provieds value for expectedHardwareConfiguration then write a function manager() which will have inputs namespace, customFilter and BMH host_list. This function will call appropriate filter written in classification_filter.go based on 'customFilter' specified by user in CRD_yaml. If user have set different expression rules instead of custom classification filters for filtering BMH then filtering will be done based on 'operation' value provided.
+
+    - The ClassificationFilters.go file will contain multiple comparision algorithms based on multiple custom classification filters(i.e minimum and maximum) or expression filters. Create a FactoryClassificationManager() function which will have inputs baremetalhost list and extracted hardware profile. FactoryClassificationManager's job is to call respective filtering algorithm and parse the inputs accordingly so respective filtering algorithm will check each host against profile and update the label like below if match found. 
+
+        ```yaml
+        "metadata": {
+            "labels": {
+                "profile.metal3.io/profile1" : "BMO-OK",
```

Comment:
It would be nice if the label name and this CRD name matched.
```suggestion
                "hardwareclassification.metal3.io/profile1" : "matches",
```

Reply:
Yes